### PR TITLE
Handle no-hash-indexed delete record in record cleaner, and fix un-deleted engine in test

### PR DESCRIPTION
**Pull request by JiayuZzz**
What's Changed:

- Fix: Engine not deleted in TestStringModify
- Cleaner: Handle old delete record of skiplist which not indexed by hash table
- Reorg PointerWithTag to make it more clear

### CODE REVIEW

File: engine/kv_engine.cpp [hunk 1]
```diff
@@ -941,7 +941,8 @@ Status KVEngine::SDeleteImpl(Skiplist* skiplist, const StringView& user_key) {
               "Wrong existing record type while insert a delete reocrd for "
               "sorted collection");
           delayFree(OldDataRecord{ret.existing_record, new_ts});
-          delayFree(OldDeleteRecord{ret.write_record, new_ts});
+          delayFree(OldDeleteRecord{ret.write_record, new_ts,
+                                    ret.hash_entry_ptr, hint.spin});
```
Comment (sherry-1001) [inline engine/kv_engine.cpp:945]: How about for `no hash index`?  its `hash_entry_ptr` is nullptr.

**Review by ZiyanShi — APPROVED** [review]

File: engine/version/old_records_cleaner.cpp [hunk 2]
```diff
@@ -207,60 +207,92 @@ SpaceEntry OldRecordsCleaner::purgeOldDataRecord(
...
 SpaceEntry OldRecordsCleaner::purgeOldDeleteRecord(
-    const OldDeleteRecord& old_delete_record) {
+    OldDeleteRecord& old_delete_record) {
   DataEntry* data_entry =
       static_cast<DataEntry*>(old_delete_record.pmem_delete_record);
   switch (data_entry->meta.type) {
     case StringDeleteRecord: {
-      if (old_delete_record.hash_entry_ref->GetIndex().string_record ==
+      kvdk_assert(
+          old_delete_record.record_index.hash_entry.RawPointer() != nullptr &&
```
Comment (sherry-1001) [inline engine/version/old_records_cleaner.cpp:216]: wrapper these check to function.

File: engine/kv_engine.cpp [hunk 3]
```diff
@@ -941,7 +941,18 @@ Status KVEngine::SDeleteImpl(Skiplist* skiplist, const StringView& user_key) {
               "Wrong existing record type while insert a delete reocrd for "
               "sorted collection");
           delayFree(OldDataRecord{ret.existing_record, new_ts});
-          delayFree(OldDeleteRecord{ret.write_record, new_ts});
+          if (ret.hash_entry_ptr != nullptr) {
```
Comment (sherry-1001) [inline engine/kv_engine.cpp:944]: adding some comments for if-else case.

Comment (sherry-1001) [reply engine/kv_engine.cpp:944]: here why not use `pointwithtag` for `ret.hash_entry_ptr`,  then check the tag.

File: engine/version/old_records_cleaner.cpp [hunk 4]
```diff
@@ -207,60 +207,92 @@ SpaceEntry OldRecordsCleaner::purgeOldDataRecord(
...
-          case HashIndexType::Empty:
-            break;
-          default:
-            GlobalLogger.Error(
-                "Wrong type %u in handle pending free skiplist delete record\n",
-                hash_entry_ref->GetIndexType());
-            std::abort();
-        }
+          } else {
+            std::abort();  // never reach
```
Comment (ZiyanShi) [inline engine/version/old_records_cleaner.cpp:257]: May be reached?

**Review by ZiyanShi — CHANGES_REQUESTED** [review]

File: engine/version/old_records_cleaner.hpp [hunk 5]
```diff
@@ -25,14 +25,33 @@ struct OldDataRecord {
 };
 
 struct OldDeleteRecord {
+  union RecordIndex {
+    RecordIndex(void* ptr, PointerType type)
```
Comment (ZiyanShi) [inline engine/version/old_records_cleaner.hpp:29]: Don't union.
PointerWithTag<void, PointerType> and use access funcions such as
Type() const
HashEntryPointer() const
SkiplistNodePointer() const
RawPointer() const
to extract the pointer according to tag.

**Review by ZiyanShi — APPROVED** [review]

File: engine/kv_engine.cpp (hunk 3 above)
Comment (JiayuZzz) [reply engine/kv_engine.cpp:944]: Will do this while optimize hash entry size to 8

File: engine/version/old_records_cleaner.hpp (hunk 5 above)
Comment (JiayuZzz) [reply engine/version/old_records_cleaner.hpp:29]: PointerWithTag<void> is not supportted

**Review by ZiyanShi — APPROVED** [review]

**Review by sherry-1001 — APPROVED** [review]

**Pull request merged**